### PR TITLE
i#3289 asm function prologue/epilogue: replace push/pop with common asm macros 

### LINKED ISSUE #3289: Add calling convention macros to asm_defines to be used in asm functions
Instead of push/pop #if/#else by architecture and system in asm code, provide proper PUSH_CALLEE_SAVED_REGS/POP_CALLEE_SAVED_REGS macros that can be used instead.

### PR DESCRIPTION
**Pull request by hgreving2304**
Add PUSH_CALLEE_SAVED_REGS/POP_CALLEE_SAVED_REGS calling convention macros that can be used by asm code independent of architecture and system.

Issue: #3289

### CODE REVIEW

**Comment by hgreving2304** [comment]
run arm tests

File: core/arch/asm_defines.asm [hunk 1]
```diff
@@ -411,6 +419,24 @@ ASSUME fs:_DATA @N@\
 #   define ARG8_NORETADDR  QWORD [7*ARG_SZ + REG_XSP]
 #   define ARG9_NORETADDR  QWORD [8*ARG_SZ + REG_XSP]
 #   define ARG10_NORETADDR QWORD [9*ARG_SZ + REG_XSP]
+#   define PUSH_CALLEE_SAVED_REGS() \
```
Comment (derekbruening) [inline core/arch/asm_defines.asm]: No definition for AARCHXX?  Won't drreg-test then fail to build on arm?

File: core/arch/asm_defines.asm [hunk 2]
```diff
@@ -458,6 +498,25 @@ ASSUME fs:_DATA @N@\
...
+        PUSH_SEH(REG_XSI) @N@\
+        PUSH_SEH(REG_XDI)
+#  define POP_CALLEE_SAVED_REGS() \
+        POP_SEH(REG_XDI) @N@\
+        POP_SEH(REG_XSI) @N@\
+        POP_SEH(REG_XBP) @N@\
+        POP_SEH(REG_XBX)
+#  else /* UNIX */
+#   define PUSH_CALLEE_SAVED_REGS() \
+         PUSH_SEH(REG_XBX) @N@\
```
Comment (derekbruening) [inline core/arch/asm_defines.asm]: UNIX should match Windows 32-bit: cdecl is all the same, with ebx, esi, edi, and ebp callee-saved, AFAIK.  Where are you seeing that esi and edi are caller-saved?

File: suite/tests/client-interface/drreg-test.c [hunk 3]
```diff
@@ -163,10 +159,7 @@ GLOBAL_LABEL(FUNCNAME:)
...
      epilog:
         add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
-        pop      REG_XDI
-        pop      REG_XSI
-        pop      REG_XBP
-        pop      REG_XBX
+        POP_CALLEE_SAVED_REGS()
         ret
 #elif defined(ARM)
         b        test1
```
Comment (derekbruening) [inline suite/tests/client-interface/drreg-test.c]: Seems like the arm side should also push and pop the callee-saved.

File: suite/tests/client-interface/drreg-test.c [hunk 4]
```diff
@@ -310,10 +292,7 @@ GLOBAL_LABEL(FUNCNAME:)
...
      epilog3:
         add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
-        pop      REG_XDI
-        pop      REG_XSI
-        pop      REG_XBP
-        pop      REG_XBX
+        POP_CALLEE_SAVED_REGS()
         ret
 #elif defined(ARM)
         b        test5
```
Comment (derekbruening) [inline suite/tests/client-interface/drreg-test.c]: What about other code: "git grep 'push callee-saved'" alone shows a number of other instances.  If you do not plan to convert them here, do not mark the issue as Fixed and mention the work still to be done in the commit message.

File: core/arch/asm_defines.asm (hunk 1 above)
Comment (hgreving2304) [reply core/arch/asm_defines.asm]: The arm version of the test for some reason doesn't emulate the ABI (no push/pop). Not sure why, and I don't have any way to test. Do you want me to add it anyway?

File: core/arch/asm_defines.asm (hunk 2 above)
Comment (hgreving2304) [reply core/arch/asm_defines.asm]: I googled multiple places, one is https://en.wikipedia.org/wiki/X86_calling_conventions
They all said System V ABI does not declare xdi, xsi callee-saved. I'll check if I find a better place of documentation

Comment (derekbruening) [reply core/arch/asm_defines.asm]: That wiki page says that for the System V **AMD64** ABI!  We're talking about 32-bit here.

File: core/arch/asm_defines.asm [hunk 5]
```diff
@@ -131,6 +131,7 @@
 # define RAW(n) .byte HEX(n) @N@
 # define DECLARE_FUNC_SEH(symbol) DECLARE_FUNC(symbol)
 # define PUSH_SEH(reg) push reg
+# define POP_SEH(reg) pop reg
```
Comment (derekbruening) [inline core/arch/asm_defines.asm]: You defined POP_SEH only for one assembler and left it undefined for the others.

Comment (hgreving2304) [reply core/arch/asm_defines.asm]: Just saw that and just updated (we do't need the pop macro) I think this was the reason for the Appveyor failure.

File: suite/tests/client-interface/drreg-test.c (hunk 3 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: I'd have to add this blindly, can't test it (except on GitHub)

File: core/arch/asm_defines.asm (hunk 2 above)
Comment (hgreving2304) [reply core/arch/asm_defines.asm]: You're right! I was making the assumption that i386 is the same. https://wiki.osdev.org/System_V_ABI#i386.
I'll fix

File: suite/tests/client-interface/drreg-test.c (hunk 4 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: I suggest I fix the occurences for x86 and leave the issue open and will add a comment that ARM/AArch64 is still TBD.

Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: They have some pseudo macros already: SAVE_PRESERVED_REGS/RESTORE_PRESERVED_REGS
They don't have the real ABI I see in llvm. I suggest to leave it open.

File: core/arch/asm_defines.asm (hunk 1 above)
Comment (derekbruening) [reply core/arch/asm_defines.asm]: Either way -- should prob have a comment `XXX i#3289` for adding it or sthg if not.

Comment (hgreving2304) [reply core/arch/asm_defines.asm]: Done

**Pull request merged**